overwatch/src/services/runner.rs

```diff
@@ -0,0 +1,350 @@
+            }
+        };
+
+        let inbound_relay = service_resources
+            .inbound_relay
+            .take()
+            .expect("Failed to retrieve inbound relay.");
+
+        let service_resources_handle = service_resources.to_handle(inbound_relay);
+        let service = Service::init(service_resources_handle, initial_state.clone());
+
+        service_resources
+            .state_updater()
+            .update(Some(initial_state));
+
+        service_resources
+            .status_handle
+            .service_runner_updater()
+            .starting();
+
+        match service {
+            Ok(service) => {
+                Self::start_tasks(
+                    service,
+                    service_resources,
+                    service_task_handle,
+                    state_handle_task_handle,
+                );
+            }
+            Err(error) => {
+                panic!("Service couldn't be initialised: {error}");
+            }
+        }
+    }
+
+    fn start_tasks<Service>(
+        service: Service,
+        service_resources: &ServiceResources<Message, Settings, State, StateOp, RuntimeServiceId>,
+        service_task_handle: &mut Option<JoinHandle<()>>,
+        state_handle_task_handle: &mut Option<JoinHandle<()>>,
+    ) where
+        Service: ServiceCore<RuntimeServiceId, Settings = Settings, State = State, Message = Message>
+            + 'static,
+        StateOp: StateOperator<State = State> + Clone,
+    {
+        let runtime = service_resources.overwatch_handle.runtime().clone();
+        let service_task = Self::create_service_run_task(service, service_resources);
+        *service_task_handle = Some(runtime.spawn(service_task));
+        let state_handle_task = service_resources.state_handle.clone().run();
+        *state_handle_task_handle = Some(runtime.spawn(state_handle_task));
+    }
+
+    fn create_service_run_task<Service>(
+        service: Service,
+        service_resources: &ServiceResources<Message, Settings, State, StateOp, RuntimeServiceId>,
+    ) -> impl Future<Output = ()>
+    where
+        Service: ServiceCore<RuntimeServiceId, Settings = Settings, State = State, Message = Message>
+            + 'static,
+        StateOp: Clone,
+    {
+        let task = service.run();
+        let lifecycle_notifier = service_resources.lifecycle_handle.notifier().clone();
+
+        // Receiver is ignored because it's pointless:
+        // - If we wait for it, the Stop message will eventually abort it before the
+        //   finished signal is received.
+        // - If we don't wait for it and the task finishes, the ServiceRunner will
+        //   ignore it.
+        let (sender, _receiver) = finished_signal::channel();
+        async move {
+            if let Err(error) = task.await {
+                error!("Error while waiting for Service's task to be completed: {error}");
+            }
+            if let Err(error) = lifecycle_notifier
```

Comment:
How do we get here? If the service already exited the loop, do we need to send the stop message?

Reply:
There's two cases the `Stop` use-case:
- User sends a `Stop`. `StatusHandle`'s task is fused, `Service`'s task is killed and then do cleanup.
- Service finishes on its own. `Service`'s task is already stopped, but we still need to fuse the `StatusHandle`'s task and do cleanup.

This would be the second use case.
I tried a couple or three different approaches, but sending a message back to the runner seemed to be the cleaner one so we could reuse code, which in turn helped, I think, with clarity.

Do you feel there's something still confusing?

Reply:
Ok, perhaps then just a comment that explains exactly what you explained here 😊